File: src/jyut-dict/logic/sentence/sourcesentence.cpp
#include "sourcesentence.h"

#include "logic/utils/chineseutils.h"

SourceSentence::SourceSentence()
{
}

SourceSentence::SourceSentence(const std::string &sourceLanguage,
                               const std::string &simplified,
                               const std::string &traditional,
                               const std::string &jyutping,
                               const std::string &pinyin,
                               const std::vector<SentenceSet> &sentences)
    : _sourceLanguage{sourceLanguage}
    , _simplified{simplified}
    , _traditional{traditional}
    , _jyutping{jyutping}
    , _pinyin{pinyin}
    , _sentences{sentences}
{
}

std::ostream &operator<<(std::ostream &out, const SourceSentence &sourceSentence)
{
    out << "Simplified: " << sourceSentence.getSimplified() << "\n";
    out << "Traditional: " << sourceSentence.getTraditional() << "\n";
    out << "Jyutping: " << sourceSentence.getJyutping() << "\n";
    out << "Pinyin: " << sourceSentence.getPinyin() << "\n";
    for (size_t i = 0; i < sourceSentence.getSentenceSets().size(); i++) {
        out << sourceSentence.getSentenceSets()[i] << "\n";
    }
    return out;
}

std::string SourceSentence::getSourceLanguage(void) const
{
    return _sourceLanguage;
}

void SourceSentence::setSourceLanguage(std::string sourceLanguage)
{
    _sourceLanguage = sourceLanguage;
}

std::string SourceSentence::getCharacters(EntryCharactersOptions options) const
{
    switch (options) {
    case EntryCharactersOptions::ONLY_SIMPLIFIED:
    case EntryCharactersOptions::PREFER_SIMPLIFIED:
        return _simplified;
    case EntryCharactersOptions::ONLY_TRADITIONAL:
    case EntryCharactersOptions::PREFER_TRADITIONAL:
        return _traditional;
    }
    return _traditional;
}

std::string SourceSentence::getSimplified(void) const
{
    return _simplified;
}

void SourceSentence::setSimplified(std::string simplified)
{
    _simplified = simplified;
}

std::string SourceSentence::getTraditional(void) const
{
    return _traditional;
}

void SourceSentence::setTraditional(std::string traditional)
{
    _traditional = traditional;
}

bool SourceSentence::generatePhonetic(CantoneseOptions cantoneseOptions,
                                      MandarinOptions mandarinOptions)
{
    if ((cantoneseOptions & CantoneseOptions::PRETTY_YALE)
            == CantoneseOptions::PRETTY_YALE
        && !_isYaleValid) {
        _yale
            = ChineseUtils::convertJyutpingToYale(_jyutping,
                                                  /* useSpacesToSegment */ true);
        _isYaleValid = true;
    }

    if ((mandarinOptions & MandarinOptions::PRETTY_PINYIN)
            == MandarinOptions::PRETTY_PINYIN && !_isPrettyPinyinValid) {
        _prettyPinyin = ChineseUtils::createPrettyPinyin(_pinyin);
        _isPrettyPinyinValid = true;
    }

    if ((mandarinOptions & MandarinOptions::NUMBERED_PINYIN)
            == MandarinOptions::NUMBERED_PINYIN && !_isNumberedPinyinValid) {
        _numberedPinyin = ChineseUtils::createNumberedPinyin(_pinyin);
        _isNumberedPinyinValid = true;
    }

    return true;

}

std::string SourceSentence::getPhonetic(EntryPhoneticOptions options,
                                        CantoneseOptions cantoneseOptions,
                                        MandarinOptions mandarinOptions) const
{
    switch (options) {
    case EntryPhoneticOptions::ONLY_CANTONESE:
    case EntryPhoneticOptions::PREFER_CANTONESE:
        return getCantonesePhonetic(cantoneseOptions);
    case EntryPhoneticOptions::ONLY_MANDARIN:
    case EntryPhoneticOptions::PREFER_MANDARIN:
        return getMandarinPhonetic(mandarinOptions);
    }
    return getCantonesePhonetic(cantoneseOptions);
}

std::string SourceSentence::getCantonesePhonetic(
    CantoneseOptions cantoneseOptions) const
{
    switch (cantoneseOptions) {
    case CantoneseOptions::PRETTY_YALE: {
        return _yale;
    }
    case CantoneseOptions::RAW_JYUTPING:
    default:
        return _jyutping;
    }
}

std::string SourceSentence::getMandarinPhonetic(
    MandarinOptions mandarinOptions) const
{
    switch (mandarinOptions) {
    case MandarinOptions::PRETTY_PINYIN:
        return _prettyPinyin;
    case MandarinOptions::NUMBERED_PINYIN:
    default:
        return _pinyin;
    }
}

std::string SourceSentence::getJyutping(void) const
{
    return _jyutping;
}

void SourceSentence::setJyutping(const std::string &jyutping)
{
    _jyutping = jyutping;
}

std::string SourceSentence::getPinyin(void) const
{
    return _pinyin;
}

std::string SourceSentence::getPrettyPinyin(void) const
{
    return _prettyPinyin;
}

void SourceSentence::setPinyin(const std::string &pinyin)
{
    _pinyin = pinyin;
}

std::vector<SentenceSet> SourceSentence::getSentenceSets(void) const
{
    return _sentences;
}

std::string SourceSentence::getSentenceSnippet(void) const
{
    if (_sentences.empty()) {
        return "";
    }

    SentenceSet sentenceSet = _sentences.at(0);

    if (sentenceSet.getSentenceSnippet().empty()) {
        return "";
    }

    std::vector<Sentence::TargetSentence> snippets = sentenceSet
                                                         .getSentenceSnippet();

    if (snippets.empty()) {
        return "";
    }

    return snippets.at(0).sentence;
}

std::string SourceSentence::getSentenceSnippetLanguage(void) const
{
    if (_sentences.empty()) {
        return "";
    }

    SentenceSet sentenceSet = _sentences.at(0);

    if (sentenceSet.getSentenceSnippet().empty()) {
        return "";
    }

    std::vector<Sentence::TargetSentence> snippets = sentenceSet
                                                         .getSentenceSnippet();

    if (snippets.empty()) {
        return "";
    }

    return snippets.at(0).language;
}

void SourceSentence::setIsWelcome(const bool isWelcome)
{
    _isWelcome = isWelcome;
}

bool SourceSentence::isWelcome(void) const
{
    return _isWelcome;
}

void SourceSentence::setIsEmpty(const bool isEmpty)
{
    _isEmpty = isEmpty;
}

bool SourceSentence::isEmpty(void) const
{
    return _isEmpty;
}

File: src/ios/CantoneseDictionary/CantoneseDictionary/logic/sentence/SourceSentence.swift
//
//  SourceSentence.swift
//  CantoneseDictionary
//
//  Created by Aaron on 2026-09-04.
//

class SourceSentence {
    private var _sourceLanguage: String
    private var _simplified: String
    private var _traditional: String

    private var _jyutping: String
    private var _yale: String?
    private var _cantoneseIPA: String?

    private var _pinyin: String
    private var _prettyPinyin: String?
    private var _numberedPinyin: String?
    private var _zhuyin: String?
    private var _mandarinIPA: String?

    private var _sentences: [SentenceSet]

    private var _isWelcome: Bool = false
    private var _isEmpty: Bool = false

    init(
        sourceLanguage: String,
        simplified: String,
        traditional: String,
        jyutping: String,
        pinyin: String,
        sentences: [SentenceSet]
    ) {
        self._sourceLanguage = sourceLanguage
        self._simplified = simplified
        self._traditional = traditional
        self._jyutping = jyutping
        self._pinyin = pinyin
        self._sentences = sentences
    }

    var sourceLanguage: String {
        get {
            _sourceLanguage
        }
        set {
            _sourceLanguage = newValue
        }
    }

    func getCharacters(options: EntryCharactersOptions) -> String {
        switch options {
        case .onlySimplified, .preferSimplified:
            return _simplified
        case .onlyTraditional, .preferTraditional:
            return _traditional
        }
    }

    var simplified: String {
        get {
            _simplified
        }
        set {
            _simplified = newValue
        }
    }

    var traditional: String {
        get {
            _traditional
        }
        set {
            _traditional = newValue
        }
    }

    func generatePhonetic(
        cantoneseOptions: CantoneseOptions,
        mandarinOptions: MandarinOptions
    ) -> Bool {
        if (cantoneseOptions.rawValue & CantoneseOptions.prettyYale.rawValue
            == CantoneseOptions.prettyYale.rawValue) && _yale == nil
        {
            // TODO: Implement
        }
        if (cantoneseOptions.rawValue & CantoneseOptions.cantoneseIPA.rawValue
            == CantoneseOptions.cantoneseIPA.rawValue) && _cantoneseIPA == nil
        {
            // TODO: Implement
        }

        if (mandarinOptions.rawValue & MandarinOptions.prettyPinyin.rawValue
            == MandarinOptions.prettyPinyin.rawValue) && _prettyPinyin == nil
        {
            // TODO: Implement
        }
        if (mandarinOptions.rawValue & MandarinOptions.numberedPinyin.rawValue
            == MandarinOptions.numberedPinyin.rawValue)
            && _numberedPinyin == nil
        {
            // TODO: Implement
        }
        if (mandarinOptions.rawValue & MandarinOptions.zhuyin.rawValue
            == MandarinOptions.zhuyin.rawValue) && _zhuyin == nil
        {
            // TODO: Implement
        }
        if (mandarinOptions.rawValue & MandarinOptions.mandarinIPA.rawValue
            == MandarinOptions.mandarinIPA.rawValue) && _mandarinIPA == nil
        {
            // TODO: Implement
        }

        return true
    }

    func getPhonetic(
        options: EntryPhoneticOptions,
        cantoneseOptions: CantoneseOptions,
        mandarinOptions: MandarinOptions
    ) -> String {
        switch options {
        case .onlyCantonese, .preferCantonese:
            return getCantonesePhonetic(cantoneseOptions: cantoneseOptions)
        case .onlyMandarin, .preferMandarin:
            return getMandarinPhonetic(mandarinOptions: mandarinOptions)
        }
    }

    func getCantonesePhonetic(cantoneseOptions: CantoneseOptions) -> String {
        switch cantoneseOptions {
        case .prettyYale:
            return _yale!
        case .cantoneseIPA:
            return _cantoneseIPA!
        case .rawJyutping:
            fallthrough
        default:
            return _jyutping
        }
    }

    func getMandarinPhonetic(mandarinOptions: MandarinOptions) -> String {
        switch mandarinOptions {
        case .prettyPinyin:
            return _prettyPinyin!
        case .numberedPinyin:
            return _numberedPinyin!
        case .zhuyin:
            return _zhuyin!
        case .mandarinIPA:
            return _mandarinIPA!
        case .rawPinyin:
            fallthrough
        default:
            return _pinyin
        }
    }

    var jyutping: String {
        get {
            _jyutping
        }
        set {
            _jyutping = newValue
        }
    }

    var pinyin: String {
        get {
            _pinyin
        }
        set {
            _pinyin = newValue
        }
    }

    func getPrettyPinyin() -> String {
        _prettyPinyin ?? ""
    }

    func getSentenceSets() -> [SentenceSet] {
        _sentences
    }

    func getSentenceSnippet() -> String {
        if _sentences.isEmpty {
            return ""
        }

        let sentenceSet: SentenceSet = _sentences[0]
        if sentenceSet.getSentenceSnippet().isEmpty {
            return ""
        }

        let snippets: [TargetSentence] = sentenceSet.getSentenceSnippet()

        if snippets.isEmpty {
            return ""
        }

        return snippets[0].sentence
    }

    func getSentenceSnippetLanguage() -> String {
        if _sentences.isEmpty {
            return ""
        }

        let sentenceSet: SentenceSet = _sentences[0]
        if sentenceSet.getSentenceSnippet().isEmpty {
            return ""
        }

        let snippets: [TargetSentence] = sentenceSet.getSentenceSnippet()

        if snippets.isEmpty {
            return ""
        }

        return snippets[0].language
    }

    var isWelcome: Bool {
        get {
            _isWelcome
        }
        set {
            _isWelcome = newValue
        }
    }

    var isEmpty: Bool {
        get {
            _isEmpty
        }
        set {
            _isEmpty = newValue
        }
    }
}
